org-code-javabuilder/lib/src/main/java/org/code/javabuilder/InputRedirectionStream.java

```diff
@@ -0,0 +1,98 @@
+package org.code.javabuilder;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedList;
+import java.util.Queue;
+
+/**
+ * An InputStream that queries an InputAdapter for new bytes. This is intended to redirect system.in
+ * to use the InputAdapter rather than the server's console. See
+ * https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html for full implementation
+ * details & contract.
+ */
+public class InputRedirectionStream extends InputStream {
+  private final Queue<Byte> queue;
+  private final InputAdapter inputAdapter;
+
+  public InputRedirectionStream(InputAdapter inputAdapter) {
+    this.inputAdapter = inputAdapter;
+    this.queue = new LinkedList<>();
+  }
+
+  /**
+   * See: https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-- Checks the queue
+   * for existing bytes. If the queue is empty, polls the inputAdapter for new data. This is a
+   * blocking call.
+   *
+   * @return the first byte in the queue
+   */
+  @Override
+  public int read() {
+    if (queue.peek() == null) {
+      byte[] message = inputAdapter.getNextMessage().getBytes(StandardCharsets.UTF_8);
+      for (byte b : message) {
+        queue.add(b);
+      }
+    }
+
+    return (int) queue.remove();
+  }
+
+  /** See: https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-byte:A- */
+  @Override
+  public int read(byte[] b) {
+    return read(b, 0, b.length);
```

Comment:
style question: when we call methods within a class, should we preface with `this.`? I personally find it easier to read because I know we are calling another method within this class, but am not sure what the preferred style is.

Reply:
That sounds good to me. I also find that easier to read.